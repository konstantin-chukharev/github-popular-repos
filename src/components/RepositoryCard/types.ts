import type { Repository } from 'src/api/types';

export type RepositoryCardOwner = Pick<
  Repository['owner'],
  'login' | 'type' | 'avatar_url' | 'html_url'
>;

export type RepositoryCardData = Pick<
  Repository,
  | 'html_url'
  | 'name'
  | 'description'
  | 'topics'
  | 'stargazers_count'
  | 'language'
  | 'created_at'
  | 'updated_at'
> & { owner: RepositoryCardOwner };

export type RepositoryCardProps = {
  repository: RepositoryCardData;
  toggleStar: () => void;
  hasStarred: () => boolean;
  isStarringAvailable: boolean;
};
