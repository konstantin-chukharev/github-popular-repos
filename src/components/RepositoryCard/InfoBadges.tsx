import { Badge, Flex, Link } from '@radix-ui/themes';
import {
  BackpackIcon,
  CodeIcon,
  PersonIcon,
  StarIcon,
} from '@radix-ui/react-icons';
import { formatDate } from 'src/utils/formatDate';

import { RepositoryCardData } from './types';

export const InfoBadges = ({
  repository,
}: {
  repository: RepositoryCardData;
}) => {
  const { owner, stargazers_count, language, created_at, updated_at } =
    repository;

  return (
    <Flex gap="1" wrap="wrap" data-testid="InfoBadges">
      <Badge
        size="2"
        color="gold"
        className="CardContentBadge"
        data-testid="StarsBadge"
      >
        <StarIcon />
        {stargazers_count}
      </Badge>

      {language ? (
        <Badge
          size="2"
          color="gray"
          className="CardContentBadge"
          data-testid="LanguageBadge"
        >
          <CodeIcon />
          <Link
            href={`https://github.com/topics/${language}`}
            color="gray"
            underline="none"
            target="_blank"
            rel="noopener noreferrer"
            data-testid="LanguageLink"
          >
            {language}
          </Link>
        </Badge>
      ) : null}

      <Badge
        size="2"
        color="gray"
        className="CardContentBadge"
        data-testid="OwnerTypeBadge"
      >
        {owner.type === 'Organization' ? <BackpackIcon /> : <PersonIcon />}
        <Link
          href={owner.html_url}
          color="gray"
          underline="none"
          target="_blank"
          rel="noopener noreferrer"
          data-testid="OwnerLink"
        >
          {owner.login}
        </Link>
      </Badge>

      <Badge
        size="2"
        color="gray"
        className="CardContentBadge"
        data-testid="CreatedAtBadge"
      >
        Created {formatDate(created_at)}
      </Badge>

      <Badge
        size="2"
        color="gray"
        className="CardContentBadge"
        data-testid="UpdatedAtBadge"
      >
        Updated {formatDate(updated_at)}
      </Badge>
    </Flex>
  );
};
