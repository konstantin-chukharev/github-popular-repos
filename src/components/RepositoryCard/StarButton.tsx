import { forwardRef } from 'react';
import { IconButton, IconButtonProps, Tooltip } from '@radix-ui/themes';
import { StarFilledIcon, StarIcon } from '@radix-ui/react-icons';

type StarButtonProps = {
  hasStarred: boolean;
  onToggleStar: () => void;
};

export const StarButton = forwardRef<
  HTMLButtonElement,
  StarButtonProps & IconButtonProps
>(({ hasStarred, disabled, onToggleStar, ...props }, ref) => {
  const content = (
    <IconButton
      {...props}
      ref={ref}
      disabled={disabled}
      variant={hasStarred ? 'solid' : 'soft'}
      onClick={onToggleStar}
      data-testid="StarButton"
    >
      {hasStarred ? <StarFilledIcon /> : <StarIcon />}
    </IconButton>
  );

  if (disabled) {
    return (
      <Tooltip
        content="Unable to start the repository. Most likely there's no access to the storage or it's full"
        data-testid="StarButtonTooltip"
      >
        {content}
      </Tooltip>
    );
  }

  return content;
});

StarButton.displayName = 'StarButton';
