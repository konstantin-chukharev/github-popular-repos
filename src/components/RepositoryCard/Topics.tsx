import { Badge, Flex, Link } from '@radix-ui/themes';

export const Topics = ({ topics }: { topics: readonly string[] }) => {
  if (!topics.length) return null;

  return (
    <Flex gap="1" wrap="wrap" data-testid="Topics">
      {topics.slice(0, 10).map((topic) => (
        <Badge key={topic} data-testid="TopicBadge">
          <Link
            color="indigo"
            underline="hover"
            target="_blank"
            rel="noopener noreferrer"
            href={`https://github.com/topics/${topic}`}
            data-testid="TopicLink"
          >
            {topic}
          </Link>
        </Badge>
      ))}
    </Flex>
  );
};
