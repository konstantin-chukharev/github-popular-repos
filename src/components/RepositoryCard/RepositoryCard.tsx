import './styles.css';

import { forwardRef } from 'react';
import { Card, Flex, Text, Link, Avatar } from '@radix-ui/themes';

import { Topics } from './Topics';
import { StarButton } from './StarButton';
import { InfoBadges } from './InfoBadges';
import type { RepositoryCardProps } from './types';

/**
 * Card displaying information about a GitHub repository.
 * Includes the repository name, owner, description, topics, language and amount of stars.
 * Allows starring and unstarring the repository.
 */
export const RepositoryCard = forwardRef<HTMLDivElement, RepositoryCardProps>(
  ({ repository, isStarringAvailable, toggleStar, hasStarred }, ref) => {
    const { owner, html_url, name, description, topics } = repository;

    return (
      <Card size="2" ref={ref} data-testid="RepositoryCard">
        <Flex gap="1" justify="between">
          <Flex gap="3" direction="column">
            <Flex gap="2" align="center">
              <Avatar
                size="2"
                width={32}
                height={32}
                src={owner.avatar_url}
                fallback={owner.login.charAt(0)}
              />

              <Link
                size="3"
                weight="bold"
                href={html_url}
                target="_blank"
                rel="noopener noreferrer"
                data-testid="RepositoryName"
              >
                {name}
              </Link>
            </Flex>

            {description ? (
              <Text
                size="2"
                weight="medium"
                color="gray"
                data-testid="RepositoryDescription"
              >
                {description}
              </Text>
            ) : null}

            <Topics topics={topics} />

            <InfoBadges repository={repository} />
          </Flex>

          <StarButton
            disabled={!isStarringAvailable}
            hasStarred={hasStarred()}
            onToggleStar={toggleStar}
          />
        </Flex>
      </Card>
    );
  },
);

RepositoryCard.displayName = 'RepositoryCard';
