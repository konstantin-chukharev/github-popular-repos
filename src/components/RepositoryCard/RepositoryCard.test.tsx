import { afterEach, describe, expect, test, vi } from 'vitest';
import { render, screen, waitFor } from 'tests/react';

import { RepositoryCard } from './RepositoryCard';
import type { RepositoryCardData, RepositoryCardProps } from './types';

describe('RepositoryCard', () => {
  const repository = {
    name: 'example-repo',
    description: 'This is an example repository',
    stargazers_count: 100,
    owner: {
      type: 'User',
      login: 'example-owner',
      avatar_url: 'https://avatars.githubusercontent.com/u/9892522?v=4',
      html_url: 'https://github.com/example-owner',
    },
    topics: ['react', 'typescript'],
    language: 'typescript',
    created_at: '2020-01-01T00:00:00',
    updated_at: '2021-01-01T00:00:00',
    html_url: 'https://github.com/example-owner/example-repo',
  } satisfies RepositoryCardData;

  const toggleStar = vi.fn(() => {});

  const renderCard = (overrideProps?: Partial<RepositoryCardProps>) =>
    render(
      <RepositoryCard
        repository={repository}
        toggleStar={toggleStar}
        hasStarred={() => false}
        isStarringAvailable
        {...overrideProps}
      />,
    );

  afterEach(() => {
    toggleStar.mockClear();
  });

  test('renders card', () => {
    renderCard();

    expect(screen.getByTestId('RepositoryCard')).toBeInTheDocument();
  });

  test('renders name', () => {
    renderCard();

    const nameLink = screen.getByTestId('RepositoryName');
    expect(nameLink).toBeInTheDocument();
    expect(nameLink).toHaveAttribute('href', repository.html_url);
  });

  test('renders description', () => {
    renderCard();

    expect(screen.getByTestId('RepositoryDescription')).toHaveTextContent(
      repository.description,
    );
  });

  test('renders stars count', () => {
    renderCard();

    expect(screen.getByTestId('StarsBadge')).toHaveTextContent(
      `${repository.stargazers_count}`,
    );
  });

  test('renders language', () => {
    renderCard();

    expect(screen.getByTestId('LanguageBadge')).toBeInTheDocument();
    expect(screen.getByTestId('LanguageLink')).toHaveTextContent(
      repository.language,
    );
    expect(screen.getByTestId('LanguageLink')).toHaveAttribute(
      'href',
      'https://github.com/topics/typescript',
    );
  });

  test('renders owner username', () => {
    renderCard();

    expect(screen.getByTestId('StarsBadge')).toHaveTextContent(
      `${repository.stargazers_count}`,
    );
  });

  test('renders topics', () => {
    renderCard();

    expect(screen.getByTestId('Topics')).toBeInTheDocument();
    expect(screen.getAllByTestId('TopicBadge')).toHaveLength(2);

    screen.getAllByTestId('TopicLink').forEach((link) => {
      expect(link).toHaveTextContent(/react|typescript/);
      expect(link).toHaveAttribute(
        'href',
        expect.stringMatching(/https:\/\/github.com\/topics\/react|typescript/),
      );
    });
  });

  test('renders star button and toggles it', async () => {
    const { rerender } = renderCard();

    const starButton = screen.getByTestId('StarButton');

    expect(starButton).toBeInTheDocument();
    expect(starButton).toHaveClass('rt-variant-soft');

    starButton.click();

    expect(toggleStar).toHaveBeenCalled();

    rerender(
      <RepositoryCard
        repository={repository}
        toggleStar={toggleStar}
        hasStarred={() => true}
        isStarringAvailable
      />,
    );

    await waitFor(() => {
      expect(starButton).toHaveClass('rt-variant-solid');
    });
  });

  test('renders disabled star button', () => {
    renderCard({ isStarringAvailable: false });

    const starButton = screen.getByTestId('StarButton');

    expect(starButton).toBeInTheDocument();
    expect(starButton).toBeDisabled();
  });

  test('matches snapshot', () => {
    renderCard();
    expect(screen.getByTestId('RepositoryCard')).toMatchSnapshot();
  });
});
