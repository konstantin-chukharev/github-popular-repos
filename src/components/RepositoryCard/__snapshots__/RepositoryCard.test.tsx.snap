// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RepositoryCard > matches snapshot 1`] = `
<div
  class="rt-reset rt-BaseCard rt-Card rt-r-size-2 rt-variant-surface"
  data-testid="RepositoryCard"
>
  <div
    class="rt-Flex rt-r-jc-space-between rt-r-gap-1"
  >
    <div
      class="rt-Flex rt-r-fd-column rt-r-gap-3"
    >
      <div
        class="rt-Flex rt-r-ai-center rt-r-gap-2"
      >
        <span
          class="rt-reset rt-AvatarRoot rt-r-size-2 rt-variant-soft"
          data-accent-color=""
        >
          <span
            class="rt-AvatarFallback"
          />
        </span>
        <a
          class="rt-Text rt-reset rt-Link rt-r-size-3 rt-r-weight-bold rt-underline-auto"
          data-accent-color=""
          data-testid="RepositoryName"
          href="https://github.com/example-owner/example-repo"
          rel="noopener noreferrer"
          target="_blank"
        >
          example-repo
        </a>
      </div>
      <span
        class="rt-Text rt-r-size-2 rt-r-weight-medium"
        data-accent-color="gray"
        data-testid="RepositoryDescription"
      >
        This is an example repository
      </span>
      <div
        class="rt-Flex rt-r-fw-wrap rt-r-gap-1"
        data-testid="Topics"
      >
        <span
          class="rt-reset rt-Badge rt-r-size-1 rt-variant-soft"
          data-accent-color=""
          data-testid="TopicBadge"
        >
          <a
            class="rt-Text rt-reset rt-Link rt-underline-hover"
            data-accent-color="indigo"
            data-testid="TopicLink"
            href="https://github.com/topics/react"
            rel="noopener noreferrer"
            target="_blank"
          >
            react
          </a>
        </span>
        <span
          class="rt-reset rt-Badge rt-r-size-1 rt-variant-soft"
          data-accent-color=""
          data-testid="TopicBadge"
        >
          <a
            class="rt-Text rt-reset rt-Link rt-underline-hover"
            data-accent-color="indigo"
            data-testid="TopicLink"
            href="https://github.com/topics/typescript"
            rel="noopener noreferrer"
            target="_blank"
          >
            typescript
          </a>
        </span>
      </div>
      <div
        class="rt-Flex rt-r-fw-wrap rt-r-gap-1"
        data-testid="InfoBadges"
      >
        <span
          class="rt-reset rt-Badge rt-r-size-2 rt-variant-soft CardContentBadge"
          data-accent-color="gold"
          data-testid="StarsBadge"
        >
          <svg
            fill="none"
            height="15"
            viewBox="0 0 15 15"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M6.97942 1.25171L6.9585 1.30199L5.58662 4.60039C5.54342 4.70426 5.44573 4.77523 5.3336 4.78422L1.7727 5.0697L1.71841 5.07405L1.38687 5.10063L1.08608 5.12475C0.820085 5.14607 0.712228 5.47802 0.914889 5.65162L1.14406 5.84793L1.39666 6.06431L1.43802 6.09974L4.15105 8.42374C4.23648 8.49692 4.2738 8.61176 4.24769 8.72118L3.41882 12.196L3.40618 12.249L3.32901 12.5725L3.25899 12.866C3.19708 13.1256 3.47945 13.3308 3.70718 13.1917L3.9647 13.0344L4.24854 12.861L4.29502 12.8326L7.34365 10.9705C7.43965 10.9119 7.5604 10.9119 7.6564 10.9705L10.705 12.8326L10.7515 12.861L11.0354 13.0344L11.2929 13.1917C11.5206 13.3308 11.803 13.1256 11.7411 12.866L11.671 12.5725L11.5939 12.249L11.5812 12.196L10.7524 8.72118C10.7263 8.61176 10.7636 8.49692 10.849 8.42374L13.562 6.09974L13.6034 6.06431L13.856 5.84793L14.0852 5.65162C14.2878 5.47802 14.18 5.14607 13.914 5.12475L13.6132 5.10063L13.2816 5.07405L13.2274 5.0697L9.66645 4.78422C9.55432 4.77523 9.45663 4.70426 9.41343 4.60039L8.04155 1.30199L8.02064 1.25171L7.89291 0.944609L7.77702 0.665992C7.67454 0.419604 7.32551 0.419604 7.22303 0.665992L7.10715 0.944609L6.97942 1.25171ZM7.50003 2.60397L6.50994 4.98442C6.32273 5.43453 5.89944 5.74207 5.41351 5.78103L2.84361 5.98705L4.8016 7.66428C5.17183 7.98142 5.33351 8.47903 5.2204 8.95321L4.62221 11.461L6.8224 10.1171C7.23842 9.86302 7.76164 9.86302 8.17766 10.1171L10.3778 11.461L9.77965 8.95321C9.66654 8.47903 9.82822 7.98142 10.1984 7.66428L12.1564 5.98705L9.58654 5.78103C9.10061 5.74207 8.67732 5.43453 8.49011 4.98442L7.50003 2.60397Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
          100
        </span>
        <span
          class="rt-reset rt-Badge rt-r-size-2 rt-variant-soft CardContentBadge"
          data-accent-color="gray"
          data-testid="LanguageBadge"
        >
          <svg
            fill="none"
            height="15"
            viewBox="0 0 15 15"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M9.96424 2.68571C10.0668 2.42931 9.94209 2.13833 9.6857 2.03577C9.4293 1.93322 9.13832 2.05792 9.03576 2.31432L5.03576 12.3143C4.9332 12.5707 5.05791 12.8617 5.3143 12.9642C5.5707 13.0668 5.86168 12.9421 5.96424 12.6857L9.96424 2.68571ZM3.85355 5.14646C4.04882 5.34172 4.04882 5.6583 3.85355 5.85356L2.20711 7.50001L3.85355 9.14646C4.04882 9.34172 4.04882 9.6583 3.85355 9.85356C3.65829 10.0488 3.34171 10.0488 3.14645 9.85356L1.14645 7.85356C0.951184 7.6583 0.951184 7.34172 1.14645 7.14646L3.14645 5.14646C3.34171 4.9512 3.65829 4.9512 3.85355 5.14646ZM11.1464 5.14646C11.3417 4.9512 11.6583 4.9512 11.8536 5.14646L13.8536 7.14646C14.0488 7.34172 14.0488 7.6583 13.8536 7.85356L11.8536 9.85356C11.6583 10.0488 11.3417 10.0488 11.1464 9.85356C10.9512 9.6583 10.9512 9.34172 11.1464 9.14646L12.7929 7.50001L11.1464 5.85356C10.9512 5.6583 10.9512 5.34172 11.1464 5.14646Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
          <a
            class="rt-Text rt-reset rt-Link rt-underline-none"
            data-accent-color="gray"
            data-testid="LanguageLink"
            href="https://github.com/topics/typescript"
            rel="noopener noreferrer"
            target="_blank"
          >
            typescript
          </a>
        </span>
        <span
          class="rt-reset rt-Badge rt-r-size-2 rt-variant-soft CardContentBadge"
          data-accent-color="gray"
          data-testid="OwnerTypeBadge"
        >
          <svg
            fill="none"
            height="15"
            viewBox="0 0 15 15"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M7.5 0.875C5.49797 0.875 3.875 2.49797 3.875 4.5C3.875 6.15288 4.98124 7.54738 6.49373 7.98351C5.2997 8.12901 4.27557 8.55134 3.50407 9.31167C2.52216 10.2794 2.02502 11.72 2.02502 13.5999C2.02502 13.8623 2.23769 14.0749 2.50002 14.0749C2.76236 14.0749 2.97502 13.8623 2.97502 13.5999C2.97502 11.8799 3.42786 10.7206 4.17091 9.9883C4.91536 9.25463 6.02674 8.87499 7.49995 8.87499C8.97317 8.87499 10.0846 9.25463 10.8291 9.98831C11.5721 10.7206 12.025 11.8799 12.025 13.5999C12.025 13.8623 12.2376 14.0749 12.5 14.0749C12.7623 14.075 12.975 13.8623 12.975 13.6C12.975 11.72 12.4778 10.2794 11.4959 9.31166C10.7244 8.55135 9.70025 8.12903 8.50625 7.98352C10.0187 7.5474 11.125 6.15289 11.125 4.5C11.125 2.49797 9.50203 0.875 7.5 0.875ZM4.825 4.5C4.825 3.02264 6.02264 1.825 7.5 1.825C8.97736 1.825 10.175 3.02264 10.175 4.5C10.175 5.97736 8.97736 7.175 7.5 7.175C6.02264 7.175 4.825 5.97736 4.825 4.5Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
          <a
            class="rt-Text rt-reset rt-Link rt-underline-none"
            data-accent-color="gray"
            data-testid="OwnerLink"
            href="https://github.com/example-owner"
            rel="noopener noreferrer"
            target="_blank"
          >
            example-owner
          </a>
        </span>
        <span
          class="rt-reset rt-Badge rt-r-size-2 rt-variant-soft CardContentBadge"
          data-accent-color="gray"
          data-testid="CreatedAtBadge"
        >
          Created 
          Jan 1, 2020
        </span>
        <span
          class="rt-reset rt-Badge rt-r-size-2 rt-variant-soft CardContentBadge"
          data-accent-color="gray"
          data-testid="UpdatedAtBadge"
        >
          Updated 
          Jan 1, 2021
        </span>
      </div>
    </div>
    <button
      class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-soft rt-IconButton"
      data-accent-color=""
      data-testid="StarButton"
    >
      <svg
        fill="none"
        height="15"
        viewBox="0 0 15 15"
        width="15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M6.97942 1.25171L6.9585 1.30199L5.58662 4.60039C5.54342 4.70426 5.44573 4.77523 5.3336 4.78422L1.7727 5.0697L1.71841 5.07405L1.38687 5.10063L1.08608 5.12475C0.820085 5.14607 0.712228 5.47802 0.914889 5.65162L1.14406 5.84793L1.39666 6.06431L1.43802 6.09974L4.15105 8.42374C4.23648 8.49692 4.2738 8.61176 4.24769 8.72118L3.41882 12.196L3.40618 12.249L3.32901 12.5725L3.25899 12.866C3.19708 13.1256 3.47945 13.3308 3.70718 13.1917L3.9647 13.0344L4.24854 12.861L4.29502 12.8326L7.34365 10.9705C7.43965 10.9119 7.5604 10.9119 7.6564 10.9705L10.705 12.8326L10.7515 12.861L11.0354 13.0344L11.2929 13.1917C11.5206 13.3308 11.803 13.1256 11.7411 12.866L11.671 12.5725L11.5939 12.249L11.5812 12.196L10.7524 8.72118C10.7263 8.61176 10.7636 8.49692 10.849 8.42374L13.562 6.09974L13.6034 6.06431L13.856 5.84793L14.0852 5.65162C14.2878 5.47802 14.18 5.14607 13.914 5.12475L13.6132 5.10063L13.2816 5.07405L13.2274 5.0697L9.66645 4.78422C9.55432 4.77523 9.45663 4.70426 9.41343 4.60039L8.04155 1.30199L8.02064 1.25171L7.89291 0.944609L7.77702 0.665992C7.67454 0.419604 7.32551 0.419604 7.22303 0.665992L7.10715 0.944609L6.97942 1.25171ZM7.50003 2.60397L6.50994 4.98442C6.32273 5.43453 5.89944 5.74207 5.41351 5.78103L2.84361 5.98705L4.8016 7.66428C5.17183 7.98142 5.33351 8.47903 5.2204 8.95321L4.62221 11.461L6.8224 10.1171C7.23842 9.86302 7.76164 9.86302 8.17766 10.1171L10.3778 11.461L9.77965 8.95321C9.66654 8.47903 9.82822 7.98142 10.1984 7.66428L12.1564 5.98705L9.58654 5.78103C9.10061 5.74207 8.67732 5.43453 8.49011 4.98442L7.50003 2.60397Z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
</div>
`;
